staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/fairqueuing.go

```diff
@@ -0,0 +1,612 @@
+*/
+
+package fairqueuing
+
+import (
+	"math"
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/clock"
+	"k8s.io/apimachinery/pkg/util/waitgroup"
+	"k8s.io/apiserver/pkg/util/flowcontrol/metrics"
+	"k8s.io/apiserver/pkg/util/shufflesharding"
+	"k8s.io/klog"
+)
+
+// queueSetFactoryImpl implements the QueueSetFactory interface
+// queueSetFactoryImpl makes QueueSet objects.
+type queueSetFactoryImpl struct {
+	// wg can be nil and is ignored in that case
+	wg waitgroup.OptionalWaitGroup
+
+	clk clock.PassiveClock
+}
+
+// NewQueueSetFactory creates a new QueueSetFactory object
+func NewQueueSetFactory(clk clock.PassiveClock, wg waitgroup.OptionalWaitGroup) QueueSetFactory {
+	return &queueSetFactoryImpl{
+		wg:  wg,
+		clk: clk,
+	}
+}
+
+// NewQueueSet creates a new QueueSet object
+// There is a new QueueSet created for each priority level.
+func (qsf queueSetFactoryImpl) NewQueueSet(name string, concurrencyLimit,
+	desiredNumQueues, queueLengthLimit, handSize int,
+	requestWaitLimit time.Duration) (QueueSet, error) {
+	return newQueueSetImpl(name, concurrencyLimit, desiredNumQueues,
+		queueLengthLimit, handSize, requestWaitLimit, qsf.clk, qsf.wg)
+}
+
+// queueSetImpl is a fair queuing implementation designed for the kube-apiserver.
+// This differs from ordinary fair queuing in three ways:
+// 1) dispatching requests to be served rather than packets to be transmitted
+// 2) serving multiple requests at once
+// 3) a request's service time is not known until it finishes
+// implementation of:
+// https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190228-priority-and-fairness.md
+type queueSetImpl struct {
+	lock                 sync.Mutex
+	name                 string
+	wg                   waitgroup.OptionalWaitGroup
+	queues               []*Queue
+	clk                  clock.PassiveClock
+	vt                   float64
+	estimatedServiceTime float64
+	lastRealTime         time.Time
+	robinIdx             int
+	// numRequestsEnqueued is the number of requests currently enqueued
+	// (eg: incremeneted on Enqueue, decremented on Dequue)
+	numRequestsEnqueued int
+	concurrencyLimit    int
+	desiredNumQueues    int
+	queueLengthLimit    int
+	requestWaitLimit    time.Duration
+	emptyHandler        EmptyHandler
+	dealer              *shufflesharding.Dealer
+}
+
+// initQueues is a helper method for initializing an array of n queues
+func initQueues(n int) []*Queue {
+	fqqueues := make([]*Queue, n, n)
+	for i := 0; i < n; i++ {
+		fqqueues[i] = &Queue{Index: i, Requests: make([]*Request, 0)}
+
+	}
+	return fqqueues
+}
+
+// newQueueSetImpl creates a new queueSetImpl from passed in parameters
+func newQueueSetImpl(name string, concurrencyLimit, desiredNumQueues, queueLengthLimit, handSize int,
+	requestWaitLimit time.Duration, clk clock.PassiveClock, wg waitgroup.OptionalWaitGroup) (*queueSetImpl, error) {
+	dealer, err := shufflesharding.NewDealer(desiredNumQueues, handSize)
+	if err != nil {
+		return nil, err
```

Comment:
How about wrapping the error, to become appropriate to the client of this function?

Reply:
Done